### PR TITLE
Components: Add Menu

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds a new `Menu` component. We do _not_ write an adapter for `MenuItem` because the new `MenuItem` is only meant to be used inside of this `Menu` component which did not previously exist. Rather than using an adapter to use the new `MenuItem`, consumers should update to just use the `Menu` component directly with its corresponding `Header` and `Item` components.

Depends on https://github.com/WordPress/gutenberg/pull/29592

## How has this been tested?
Run storybook (`npm run storybook:dev`), go to the experimental components `Menu` and test that the menu item is selectable and appears correctly. Also test with keyboard navigation to try to focus the menu item and select it using the keyboard (`Space` or `Enter`)

## Screenshots 

Example of the Menu item

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/ui/menu/component.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * </Menu>
+ * ```
+ *
+ * @see https://reakit.io/docs/menu/#menu
+ *
+ * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
+ * @param {import('react').Ref<any>} forwardedRef
+ */
+function Menu( props, forwardedRef ) {
+	const { children, className, menu, ...otherProps } = useContextSystem(
```
Comment (diegohaz) [inline packages/components/src/ui/menu/component.js]: It looks like this `menu` is `undefined`, so `View` is used instead of `ReakitMenu`. `View` isn't a tabbable element. That's why it's not accessible with the keyboard.

How can we inject `menu` (which I believe would come from `useMenuState`, right?) into the context system?

File: packages/components/src/ui/menu/component.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		props,
+		'Menu'
+	);
+
+	const resizeListener = usePopoverResizeUpdater( {
+		onResize: menu?.unstable_update,
```
Comment (diegohaz) [inline packages/components/src/ui/menu/component.js]: If we're going to use unstable features from packages, we should make sure we [use the exact version for that package on package.json](https://reakit.io/docs/experimental/#use-exact-version-in-packagejson) so as to avoid issues like #23332.

**Review by diegohaz — COMMENTED** [review]
Thanks for working on this @sarayourfriend! ❤️ 

What's the reason for the fallback to `View` when there's no `menu` prop? In which case `menu` will exist and not?

Is this component always intended to be used as a dropdown menu (with a menu button that opens a popover)? I ask this because Reakit's Menu component can only be used like that. Maybe that's the reason for the fallback?

**Comment by sarayourfriend** [comment]
> Is this component always intended to be used as a dropdown menu

I'm not sure, @ItsJonQ will have to answer that, but I think your hunch about that being the reason for the fallback makes sense as there is a separate `Dropdown` component in the new component system.

**Comment by ItsJonQ** [comment]
> I'm not sure, @ItsJonQ will have to answer that, but I think your hunch about that being the reason for the fallback makes sense as there is a separate Dropdown component in the new component system.

Haiii!! @sarayourfriend  + @diegohaz I can provide some context :).

I created the `Menu` component to be able to be used in a dropdown or in a sidebar menu (similar to the Material [List](https://material-ui.com/components/lists/)).

But it sounds like the Reakit menu is intentionally designed to only work within a `Dropdown`/`Popover`-like UI, correct?

If so, we may need to create a dedicated component for Sidebar/navigation based menus 🤔

**Comment by sarayourfriend** [comment]
> If so, we may need to create a dedicated component for Sidebar/navigation based menus 🤔

I think we could solve this by adding tab indexes right? Or are there specific aria roles/attributes we need to apply to a sidebar type menu?

**Comment by ItsJonQ** [comment]
> I think we could solve this by adding tab indexes right? Or are there specific aria roles/attributes we need to apply to a sidebar type menu?

I think it's just a tab index change. @diegohaz What do you think :).

Oh! It's worth noting that I previously experienced focus/scroll jumping when I used the Reakit `useMenu` within a sidebar context. Likely due to the fact that we shouldn't be using this outside of Dropdowns, haha.

**Comment by diegohaz** [comment]
> > I think we could solve this by adding tab indexes right? Or are there specific aria roles/attributes we need to apply to a sidebar type menu?
> 
> I think it's just a tab index change. @diegohaz What do you think :).

A menu that's always visible is a `menubar`. From https://www.w3.org/TR/wai-aria-1.1/#menubar:

> A presentation of menu that usually remains visible and is usually presented horizontally.

But since we're talking about adding tab indexes, I guess we should use no ARIA in this case. For a static list of actions, just use a static list.

**Comment by sarayourfriend** [comment]
Okay I tried adding `tabIndex`es to the component when appropriate... I also added an `onSelect` prop to the MenuItem to abstract away the need for passing both `onClick` and `onKeyDown`... what do y'all think about that @diegohaz and @ItsJonQ ?

File: packages/components/src/ui/menu/component.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ * </Menu>
+ * ```
+ *
+ * @see https://reakit.io/docs/menu/#menu
+ *
+ * @param {import('@wp-g2/create-styles').ViewOwnProps<import('./types').Props, 'div'>} props
+ * @param {import('react').Ref<any>} forwardedRef
+ */
+function Menu( props, forwardedRef ) {
+	const { as, children, className, menu, ...otherProps } = useContextSystem(
```
Comment (diegohaz) [inline packages/components/src/ui/menu/component.js]: I still don't understand when this `menu` prop will be populated. Will `useMenuState` be called somewhere? cc @ItsJonQ

Comment (sarayourfriend) [reply packages/components/src/ui/menu/component.js]: It is called in `Dropdown`: 

https://github.com/ItsJonQ/g2/blob/main/packages/components/src/Dropdown/Dropdown.js#L29

Then passed to `Menu` via `DropdownMenu`: 

https://github.com/ItsJonQ/g2/blob/main/packages/components/src/Dropdown/DropdownMenu.js#L25

File: packages/components/src/ui/menu/component.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	const classes = cx( styles.Menu, className );
+	const menuProps = menu || {};
+	const Component = as || menu ? ReakitMenu : View;
+
+	return (
+		<MenuContext.Provider value={ contextProps }>
+			{ /* @ts-ignore TS doesn't like the union of PolymorphicComponent and Reakit Component but they are indeed renderable */ }
+			<Component
+				hideOnClickOutside={ false }
```
Comment (diegohaz) [inline packages/components/src/ui/menu/component.js]: Why this is set to `false`?

File: packages/components/src/ui/menu/component.js (hunk 3 above)
Comment (diegohaz) [reply packages/components/src/ui/menu/component.js]: So we would use it like this or `DropdownMenu` will be part of `@wordpress/components`?

```jsx
import { DropdownMenu } from '@wp-g2/components';
import { MenuItem } from '@wordpress/components';

<DropdownMenu>
  <MenuItem />
  <MenuItem />
</DropdownMenu>
```

Can we add stories for that?

**Comment by diegohaz** [comment]
> Okay I tried adding `tabIndex`es to the component when appropriate... I also added an `onSelect` prop to the MenuItem to abstract away the need for passing both `onClick` and `onKeyDown`... what do y'all think about that @diegohaz and @ItsJonQ ?

I think we should just render a `<button>` instead of adding `tabIndex` and working around the click behavior. Is this possible?

File: packages/components/src/ui/menu/component.js (hunk 3 above)
Comment (sarayourfriend) [reply packages/components/src/ui/menu/component.js]: I was planning on adding stories for it when I opened the next PR in this series to add the `DropdownMenu` component. We can't add them yet because the context they use will be different due to the different packages.

**Comment by sarayourfriend** [comment]
> I think we should just render a <button> instead of adding tabIndex and working around the click behavior. Is this possible?

You know I thought we were, and indeed we do render a `BaseButton` but pass `as={ View }` when the `menu` context doesn't exist... it should be easy to just pass `button` there instead of `View`.

File: packages/components/src/ui/menu/component.js (hunk 3 above)
Comment (diegohaz) [reply packages/components/src/ui/menu/component.js]: Got it! Does `DropdownMenu` depend on this PR? I feel like this PR can't be properly reviewed without `DropdownMenu`.

Comment (sarayourfriend) [reply packages/components/src/ui/menu/component.js]: Yeah, `DropdownMenu` does depend on this. I wanted to add them separately otherwise it was too big of a PR but I'm happy to add it in here if you think that's overall better.

Comment (diegohaz) [reply packages/components/src/ui/menu/component.js]: I see! I agree that it's better to have a separate PR for `DropdownMenu`. I'd suggest then removing everything that can only be used with `DropdownMenu` from the component in this PR (including the changes related to Reakit), and then cherry-pick it on the `DropdownMenu` PR. What do you think?

Comment (sarayourfriend) [reply packages/components/src/ui/menu/component.js]: Sure, I can remove the menu aware stuff from this PR.

File: packages/components/src/ui/menu/component.js (hunk 4 above)
Comment (sarayourfriend) [reply packages/components/src/ui/menu/component.js]: @ItsJonQ ?

Comment (ItsJonQ) [reply packages/components/src/ui/menu/component.js]: Haii!!! If memory serves me correctly... I believe this was to accommodate this component being used either in a Sidebar-like context vs. a Dropdown. For the Sidebar use case, the menu needed to always be visible

**Comment by ItsJonQ** [comment]
> it should be easy to just pass button there instead of View
@sarayourfriend Looks like that worked :D

For your `Remove Dropdown awareness from Menu and MenuItem` update...
I'm guessing when we add Dropdown, we'll manually rig the `reakit/Menu` code into these new `Menu` components? If so, that's a 👍 from me!

---

(👇 Scroll to bottom for TLDR)

### Original Design

This discussion and breakdown was really insightful for me.
(Forgive me for repeating info) The idea behind the original `Dropdown` and `Menu` components were for:

1. Menu = List of actions/navigation items. General Purpose.
2. Dropdown = General trigger/popover combo, using `Menu` for actions/navigation.

Even with that in mind, I believe I may have unnecessarily tangled the implementation - that being making Menu "aware" of Dropdown (which @sarayourfriend untangled [here](https://github.com/WordPress/gutenberg/pull/29645/commits/69cf935ecc605c479362e58d82fb0da2e557ee31)).

### Naming Conventions

This PR nudged me to look more into the name "Dropdown" vs. "Menu". Unfortunately, it's still unclear 😅 .
Sometimes `Menu` is a dropdown... sometimes it isn't (depends on the library).

| Component | Description | Other Names |
| --- | --- | --- |
| Menu | Collection of actionable items | Sidebar Menu, List, List Group, Nav Menu |
| Dropdown | Trigger + Popover with a collection of actionable items | Menu, Dropdown Menu, Select, Listbox |

In the case of Gutenberg and `@wordpress/components`, "Dropdown" seems to be the preferred/common name for the UI that combines a trigger + popover + actions.

With all that said...

---

### TLDR

Ultimately, we need 2 components:

1. General list of links/actions (aka. Menu)
2. Trigger + Popover with a list of links/actions (aka. Dropdown)

We may re-use the Menu UI within Dropdown (and friends), but we don't have to.
They can be separate things.

I think this PR provides us with an opportunity to intentionally define how we want these base level components to be :)

(Doesn't have to be in this PR)

---

Hope this helps! Love to hear your thoughts ❤️

**Comment by sarayourfriend** [comment]
I like your summary of the state of things @ItsJonQ. Worst case we can employ these component's polymorphism to reuse them if desired without muddling up their internals.

**Comment by diegohaz** [comment]
We already have standard `role="menu"` and `role="menuitem"` attributes in HTML, and there are ARIA patterns describing they should behave as a dropdown menu with focus managed via the roving tab index or aria-activedescendant methods. The kind of menu that's not in a dropdown and is always visible — but still manages focus as a menu — is a `role="menubar"`. That's why Reakit uses the names `Menu`, `MenuItem` and `MenuBar` for the equivalent components.

Even though these behaviors are not required, I don't think we should use these roles in a static list with tabbable elements like this component, which would be better described as `role="list"` and `role="listitem"` (or just `ul` and `li`).

Using the names `Menu` and `MenuItem` here is confusing, especially as we get more used to the ARIA patterns. I understand that they make sense on the styling aspect and `List` and `ListItem` would be too generic and wouldn't communicate their design properly. But I'd consider naming it differently, like `ActionList`, `StaticMenu`, or something else that doesn't cause conflicts with existing patterns.

On the API aspect, I prefer separate components for different semantics. For example, instead of `MenuItem` being aware of `Dropdown`, I'd rather have a `DropdownItem` component.

Alternatively, I've seen other libraries exposing very generic components like `Item` that could be used like so:

```jsx
<List>
  <Item />
  <Item />
  <Item />
</List>
```
```jsx
<ActionList>
  <Item />
  <Item />
  <Item />
</ActionList>
```
```jsx
<Menu>
  <Item />
  <Item />
  <Item />
</Menu>
```
```jsx
<DropdownMenu>
  <Item />
  <Item />
  <Item />
</DropdownMenu>
```

But, instead of `Item` being aware of all these other container components, it would be really extensible, and there would be an `ItemContext` that these other components could use to inject props and change the `Item` semantics.

**Comment by sarayourfriend** [comment]
Major +1 to Diego's solution to this problem.

**Comment by alexstine** [comment]
I tried testing this, but the G2 components page makes it very hard. I selected Menu and then found the iframe in whhich the content is loaded. When I tried to select the button inside the iframe, nothing happened. The "Code is Poetry" button. It did not appear to toggle aria-expanded either. Maybe I'm looking in the wrong place?

**Comment by sarayourfriend** [comment]
@alexstine As long as you're inside of Storybook you're looking at the right place. We're going to close this PR in favor of #30097 actually so if you have time, testing that PR instead would be better... but it's still in Storybook 😞
